Update incorrect docstring of various layers in block.py

visionNoob opened the pull request:
<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
--->

# Related Issue
- This PR closes #18585, This PR enhances documentation clarity for "almost al" layers in `block.py` without any existing linked issues.

# Overview 
This PR updates the docstrings of various layers (e.g., `SPPF`, `Bottleneck`, `C3`) in `block.py` to better reflect their functionality, parameters, and input/output expectations. The goal is to reduce ambiguity and improve developer experience when integrating these layers into custom models.
# Main Changes
### 1. Docstring Updates for ALMOST All Layers
- Key Corrections:
    - ex1) Fixed incorrect expansion of `lk` from "local key" → "large kernel"
    - ex2) `CIB Block`: Renamed from "`Conditional Identity Block`" → "`Compact Inverted Block`" to reflect its architecture.

- Parameter Clarifications:

### 2. LLM-Generated Error Mitigation
- Removed nonsensical phrases (e.g., "local key aggregation") and replaced them with mathematically grounded descriptions.


# Additional Context
- **No Logic Changes**: This PR focuses solely on documentation; no code behavior was modified.
- **Impact**: Clearer docstrings reduce debugging time and help developers use layers correctly in custom architectures.


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR significantly improves the documentation and clarity of Ultralytics' core neural network modules by adding detailed docstrings, usage examples, and clearer explanations for each class and method.

### 📊 Key Changes
- 📝 Added comprehensive docstrings to all major model blocks (e.g., C2f, C3, SPP, Bottleneck, Attention, etc.), including:
  - Clear descriptions of each module's purpose and architecture.
  - Argument and attribute explanations.
  - Example usage code for each module.
  - Output shapes and expected behaviors.
- 🧑‍💻 Improved inline comments and clarified function/method behaviors.
- 🧩 Enhanced readability and maintainability for both developers and users.
- 🧪 Provided practical code snippets to demonstrate how to instantiate and use each module.

### 🎯 Purpose & Impact
- 📚 **Better Developer Experience:** Makes it much easier for developers to understand, use, and extend Ultralytics model components.
- 🤝 **Lower Learning Curve:** Non-expert users and new contributors can now quickly grasp what each module does and how to use it.
- 🛠️ **Easier Maintenance:** Clearer code and documentation reduce onboarding time and help prevent bugs or misuse.
- 🚀 **Community Growth:** Encourages more open-source contributions and experimentation by making the codebase more accessible.

Overall, this PR transforms the code into a much more user-friendly and developer-friendly resource, benefiting the entire Ultralytics and YOLO community! 🎉📖

visionNoob committed:
Update docstrings for block.py (#18585)

visionNoob committed:
update docstrings for block.py cont (#18585)

visionNoob committed:
Merge commit '3d4a86d2ee35e52b99a65f35126f758078fcccac' into docs/fix-blocks-docstring

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

UltralyticsAssistant commented:
👋 Hello @visionNoob, thank you for your meticulously detailed pull request on updating docstrings in `block.py`! 🚀 It's fantastic to see contributions aimed at enhancing documentation clarity and improving the developer experience 👏.

To proceed with the integration of this PR, please ensure the following checklist is met:

- ✅ **Define a Purpose**: Clearly link your PR to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). While you've mentioned that this PR closes #18585, ensure the issue is properly updated with this context. Your updates seem thorough – great work! Please double-check that your PR summary accurately reflects the scope of your changes.
- ✅ **Synchronize with Source**: Confirm your PR is up-to-date with the `ultralytics/ultralytics` `main` branch. If it’s behind, click the ‘Update branch’ button or run `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any pipeline tests fail, please address them accordingly.
- ✅ **Update Documentation**: While this PR enhances docstrings in the `block.py` file, ensure any broader usage affected by your updates (e.g., tutorials or guides) is also aligned with the improved documentation. You can refer to our official [documentation guide](https://docs.ultralytics.com/).
- ✅ **Add Tests (If Applicable)**: If applicable, test updates would ensure your enhancements don't unintentionally break other components. If only documentation is changed, this step may not apply but do confirm it's strictly docstring modifications.
- ✅ **Sign the CLA**: If this is your first contribution to Ultralytics, ensure you've signed the Contributor License Agreement (CLA). You can do so by visiting our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) and commenting, “I have read the CLA Document and I sign the CLA” if you haven’t already done so.
- ✅ **Minimize Changes**: Your PR description specifies that this is a **documentation-only update** with no logic changes. Be sure to confirm that the code diff only targets docstrings and does not affect the underlying code logic. In the case of any inadvertent code changes, please revert those.

For additional context and guidance, you may find our [Contributing Guide](https://docs.ultralytics.com/help/contributing/) helpful. 

🔍 If there's anything you think could benefit from refinement (e.g., phrasing, parameter descriptions, example clarity), feel free to let us know or update your PR accordingly.

⚡ **Quick Note**: This response is automated 🤖, but an Ultralytics engineer will review your PR in detail shortly and provide additional feedback or approval. Thank you again for enhancing our codebase and contributing to the community! 💪

UltralyticsAssistant committed:
Merge branch 'main' into docs/fix-blocks-docstring

glenn-jocher commented:
@visionNoob please see the failing Docs check for issues. Thank you!

visionNoob committed:
Merge commit '748c380fc5607400deff2575ca95c1b30a7607bc' into docs/fix-blocks-docstring

visionNoob committed:
Fix Docs Returns

visionNoob committed:
Remove Args Docs from ContrastiveHead cuz its none

visionNoob committed:
Fix reference in Docs description

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

visionNoob committed:
Fix CBFuse Reference

visionNoob committed:
Merge commit '3985c2199fe083681006c1f7b3fd0d9f2b2c5364' into docs/fix-blocks-docstring

visionNoob committed:
Fix docs: Resolve MkDocs autorefs issue with numeric reference

visionNoob committed:
Updated docstring to prevent incorrect linking by escaping or clarifying

visionNoob committed:
Revert unintended code changes while updating docs.

- Restored code to its original state after inadvertent modifications during docstring updates.
- Only docstring was intended to be changed, and code is now reverted to its correct version.

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

visionNoob committed:
Fix docstring: Correct the Returns section formatting.

- Adjusted the Returns section in the docstring to follow proper format: (torch.Tensor).

visionNoob committed:
Merge commit 'cd51de7ff3fdebfdb6a9d31609db6a23df2fa628' into docs/fix-blocks-docstring

visionNoob commented:
@glenn-jocher  I have updated the docstring as requested and resolved the MkDocs autorefs issues.  All checks have passed successfully. Please review the changes, thank you!

glenn-jocher commented:
Thank you for the update and resolving the MkDocs issues! The changes look good, and we'll review them promptly.

glenn-jocher committed:
Merge branch 'main' into docs/fix-blocks-docstring

glenn-jocher commented on ultralytics/nn/modules/block.py:
```diff
@@ -1049,31 +2415,56 @@ class C2fPSA(C2f):
         super().__init__(c1, c2, n=n, e=e)
         self.m = nn.ModuleList(PSABlock(self.c, attn_ratio=0.5, num_heads=self.c // 64) for _ in range(n))
 
+    def forward(self, x):
+        """
+        Forward pass through the C2fPSA module.
+
+        Args:
+            x (torch.Tensor): Input tensor of shape (B, c1, H, W).
+
+        Returns:
+            (torch.Tensor): Output tensor of shape (B, c2, H, W) maintaining spatial dimensions.
+        """
+        return super().forward(x)
+
```
We don't want to add additional methods here. I think the main benefit/purpose of basing classes is so that you don't need to recreate all the methods in the base class.

visionNoob replied:
@glenn-jocher  Thanks for the review. I've removed the unnecessary `forward` override, relying instead on the base class implementation as advised.

fazrigading commented:
Is there any resolution for this pull request? I think it should be accepted ASAP cause I have seen a lot incorrect actual block name (like PSA = Position Sensitive Attention) on paper like https://arxiv.org/pdf/2501.13400 and https://www.analyticsvidhya.com/blog/2024/10/yolov11-object-detection/ because the docstrings are generated mostly by LLM.

visionNoob committed:
refactor: Remove unnecessary forward method from C2fPSA module

Removed redundant `forward` method override in `C2fPSA`, relying instead on inheritance from base class to simplify implementation as suggested by @glenn-jocher.

fazrigading commented:
> @fazrigading
> 
> > Is there any resolution for this pull request? I think it should be accepted ASAP cause I have seen a lot incorrect actual block name (like PSA = Position Sensitive Attention) on paper like https://arxiv.org/pdf/2501.13400 and https://www.analyticsvidhya.com/blog/2024/10/yolov11-object-detection/ because the docstrings are generated mostly by LLM.
> 
> Sorry for the delayed resolution 😭 I just resolved the issue and updated the PR accordingly. I completely agree with you

Let's not forget about this

glenn-jocher commented:
@fazrigading @visionNoob you're absolutely right about the importance of accurate documentation, especially given its impact on academic work and community understanding. We'll prioritize reviewing and merging this PR to ensure the correct information is available. Thanks for the persistence on this important fix!

Laughing-q committed:
fix conflicts

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q committed:
Merge branch 'main' into docs/fix-blocks-docstring

Laughing-q committed:
Update modules/block.py

Laughing-q committed:
Update modules/block.py

Laughing-q committed:
Update modules/block.py

Laughing-q committed:
Update modules/block.py

Laughing-q committed:
Update modules/block.py

Laughing-q committed:
Update engine/results.py

Laughing-q committed:
Update modules/block.py

UltralyticsAssistant committed:
Merge branch 'main' into docs/fix-blocks-docstring

glenn-jocher commented:
Thank you for catching the wrong block names. The key fixes are now in main: CIB reads Compact Inverted Block and lk reads large kernel in block.py. The rest is more verbose than our concise style and the branch now conflicts, so we will close this. Focused follow-up PRs are very welcome. Thanks again for contributing to Ultralytics YOLO — please don't hesitate to open a new PR if you have other ideas, we'd love to see them! 🚀
